kernel/src/vm/vmar/mod.rs

```diff
@@ -1,928 +1,18 @@
-        // Here, we use a simple brute-force FIRST-FIT algorithm.
-        // Allocate as low as possible to reduce fragmentation.
-        let mut last_end: Vaddr = VMAR_LOWEST_ADDR;
-        for vm_mapping in self.vm_mappings.iter() {
-            let range = vm_mapping.range();
-
-            debug_assert!(range.start >= last_end);
-            debug_assert!(range.end <= highest_occupied);
-
-            let last_aligned = last_end.align_up(align);
-            let needed_end = last_aligned
-                .checked_add(size)
-                .ok_or(Error::new(Errno::ENOMEM))?;
-
-            if needed_end <= range.start {
-                return Ok(last_aligned..needed_end);
-            }
-
-            last_end = range.end;
-        }
-
-        return_errno_with_message!(Errno::ENOMEM, "Cannot find free region for mapping");
-    }
-
-    /// Splits and unmaps the found mapping if the new size is smaller.
-    /// Enlarges the last mapping if the new size is larger.
-    fn resize_mapping(
-        &mut self,
-        vm_space: &VmSpace,
-        map_addr: Vaddr,
-        old_size: usize,
-        new_size: usize,
-        rss_delta: &mut RssDelta,
-    ) -> Result<()> {
-        debug_assert_eq!(map_addr % PAGE_SIZE, 0);
-        debug_assert_eq!(old_size % PAGE_SIZE, 0);
-        debug_assert_eq!(new_size % PAGE_SIZE, 0);
-
-        if new_size == 0 {
-            return_errno_with_message!(Errno::EINVAL, "cannot resize a mapping to 0 size");
-        }
-
-        if new_size == old_size {
-            return Ok(());
-        }
-
-        let old_map_end = map_addr.checked_add(old_size).ok_or(Errno::EINVAL)?;
-        let new_map_end = map_addr.checked_add(new_size).ok_or(Errno::EINVAL)?;
-        if !is_userspace_vaddr(new_map_end - 1) {
-            return_errno_with_message!(Errno::EINVAL, "resize to an invalid new size");
-        }
-
-        if new_size < old_size {
-            self.alloc_free_region_exact_truncate(
-                vm_space,
-                new_map_end,
-                old_map_end - new_map_end,
-                rss_delta,
-            )?;
-            return Ok(());
-        }
-
-        self.alloc_free_region_exact(old_map_end, new_map_end - old_map_end)?;
-
-        let last_mapping = self.vm_mappings.find_one(&(old_map_end - 1)).unwrap();
-        let last_mapping_addr = last_mapping.map_to_addr();
-        debug_assert_eq!(last_mapping.map_end(), old_map_end);
-
-        self.check_extra_size_fits_rlimit(new_map_end - old_map_end)?;
-        let last_mapping = self.remove(&last_mapping_addr).unwrap();
-        let last_mapping = last_mapping.enlarge(new_map_end - old_map_end);
-        self.insert_try_merge(last_mapping);
-        Ok(())
-    }
-}
+use ostd::mm::{MAX_USERSPACE_VADDR, Vaddr};
+use vmar_impls::RssDelta;
+pub use vmar_impls::{RssType, Vmar};
 
 pub const VMAR_LOWEST_ADDR: Vaddr = 0x001_0000; // 64 KiB is the Linux configurable default
 const VMAR_CAP_ADDR: Vaddr = MAX_USERSPACE_VADDR;
```

Comment:
Since `MAX_USERSPACE_VADDR` is only used once, consider using `ostd::mm::MAX_USERSPACE_VADDR` instead.

Mixing `pub` imports, imports for the current file, and imports for sub-modules in `mod.rs` is always painful to read and maintain. Therefore, I'd prefer to
1. avoid any imports for sub-modules (like the comment above) and
2. minimize imports for the current file if possible (like this comment).